### DISCUSSION TITLE
FuncDataTemplate<T> issue with usercontrols

### DISCUSSION THREAD

**Discussion by tiborkerenyi** [AvaloniaUI/Avalonia#16724] [Q&A] [answered]
### Describe the bug

On the site https://docs.avaloniaui.net/docs/concepts/templates/creating-data-templates-in-code described code snippet doesn't work with usercontrol.

### To Reproduce

Replace TextBlock with a usercontrol in the code, and check in usercontrol's codebehind the property's value. It will be always null.
The code throws no error.

### Expected behavior

Same behaviour as normal controls.

### Avalonia version

11.1.1

### OS

Windows

**Comment by stevemonaco** [comment]
Please provide your `UserControl` implementation with the property.

**Comment by tiborkerenyi** [comment]
Here you are. :)

axaml:
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="SchFeladatutemezo.UserControls.ucFoNaptarStack">

	<StackPanel Name="Panel" Orientation="Vertical" MinWidth="100" MinHeight="75">

	</StackPanel>
	
</UserControl>

axaml.cs:

using Avalonia;
using Avalonia.Controls;
using Avalonia.Data;
using Avalonia.Markup.Xaml;
using Avalonia.Media;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;


namespace SchFeladatutemezo.UserControls;

public partial class ucFoNaptarStack : UserControl
{
    public ucFoNaptarStack()
    {
        InitializeComponent();

        var t = Tag;
       
       
        int i = 0;  //Breakpoint
    }
}

**Comment by stevemonaco** [comment]
I'm sorry but that doesn't demonstrate anything binding-related, so I can't determine what the issue is. Can you create a full repro and either upload the zipped project or link a GitHub repo?

**Comment by tiborkerenyi** [comment]
Of course and thanks.

It runs on .NET 8.0, WIN2019 Server, VS2022

[SchFeladatutemezo.zip](https://github.com/user-attachments/files/16650669/SchFeladatutemezo.zip)

**Comment by stevemonaco** [comment]
Ok, I see. You expect `Tag` to be available in the constructor on this line:

```cs
public ucFoNaptarStack()
{
    InitializeComponent();

    var t = Tag;
}
```

From this binding:

```cs
foreach (var ember in EmberekLista)
{
    var template = new FuncDataTemplate<FoNaptarSor>((value, namescope) =>
        new ucFoNaptarStack
        {   
            [!ucFoNaptarStack.TagProperty] = new Binding("FoNaptarFeladatblokkok")
                    
        });


    FoNaptar.Columns.Add(new DataGridTemplateColumn()
    {
        Header = ember.Nev + Environment.NewLine + ember.Rendszam,
        Width = new DataGridLength(200, DataGridLengthUnitType.Pixel),
        CellTemplate = template,
        Tag = new Binding("SID")
    });
}
```

So this issue is your expectations of when data is available. Controls that do not create and assign their own `DataContext` in the constructor (or in their XAML) will not have any data binded properties available at that time. The `DataContext` is created and assigned post-construction. You can check DevTools to see that `Tag` does have the binding established at runtime:

![image](https://github.com/user-attachments/assets/6e42566a-da85-46d1-9019-66759fb6b32f)

How much later depends on the external code. It may be available when the `Loaded` / `AttachedToVisualTree` events occur, ~~but you may need to wait until `DataContextChanged` (which would be the safest).~~

**Comment by stevemonaco** [comment, marked as answer]
```cs
protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
{
    base.OnAttachedToVisualTree(e);
    var t = Tag;
}
```

This appears to work better than overriding `OnDataContextChanged` as the property bindings seem to not be established at that point. I guess `OnDataContextEndUpdate` would the `DataContext`-related event you'd want.

You could also subscribe to a property change:

```cs
public ucFoNaptarStack()
{
    InitializeComponent();
    TagProperty.Changed.Subscribe(OnTagChanged);
}

private void OnTagChanged(AvaloniaPropertyChangedEventArgs<object?> args)
{
    var t = args.NewValue;
}
```

**Comment by tiborkerenyi** [comment]
Thanks a lot! It works like a charm! :)

You are the best support person I have ever spoken to!

**Comment by tiborkerenyi** [comment]
May I have one more question?

Because of the performance of the sorting I changed my strategy, Now I have a DataTable, wich is built from strings:

        NaptarDataTable = new();
        NaptarDataTable.Clear();

        //Dátum oszlop hozzáadása
        NaptarDataTable.Columns.Add("Dátum", typeof(string));
        
        //Munkatársak hozzáadása
        foreach (var e in EmberekLista)
        {
            NaptarDataTable.Columns.Add((e.Nev + Environment.NewLine + e.Rendszam).ToString(), typeof(string));
        }

        //Sorok feltöltése
        for (int i = 0; i < 300; i++)
        {
            string date = DateTime.Now.AddDays(i).Date.ToString("yyyy. MM. dd");

            DataRow dr = NaptarDataTable.NewRow();
            dr[0] = date;
            int j = 1;
            foreach (var e in EmberekLista)
            {
                dr[j] = e.SID + "|" + date;
                j++;
            }
            NaptarDataTable.Rows.Add(dr);
        }

When I'm building my DataGrid dynamically, the regular DataGridTextColumn works fine with entire DataTable, but the DataGridTemplateColumns fail. The error message is about "Missing template", but shows no specific. I've tried many types (string, string[], DataRow, DataRowCollection, etc.), but had no match. 

The FoNaptar.Columns.Add(column); caused the error.


Have you got any idea?

The critical code is: 

        foreach (System.Data.DataColumn col in (DataContext as MainViewModel).NaptarDataTable.Columns)
        {
            if (col.Ordinal < 1)
            {
                FoNaptar.Columns.Add(new DataGridTextColumn()
                {
                    Header = col.ColumnName,
                    Binding = new Avalonia.Data.Binding($"Row.ItemArray[{col.Ordinal}]")
                });
            }

            else
            {   
                var column = new DataGridTemplateColumn()
                {
                    Header = col.ColumnName,
                    IsReadOnly = false,
                    CellEditingTemplate = new FuncDataTemplate<string>((_, _) => new TextBlock
                    {
                        [!TextBlock.TextProperty] = new Avalonia.Data.Binding($"Row.ItemArray[{col.Ordinal}]"),

                    })
                };
                FoNaptar.Columns.Add(column);
            }

**Reply by stevemonaco** [reply]
You probably want to open a separate discussion on that. I'm not very familiar with the area (or `DataTable`).

If I had to guess, it's because there's no `CellTemplate`. Keep in mind that `CellEditingTemplate` is the control shown when editing has been initiated. Otherwise, the `DataGrid` could be forced to show hundreds of much more expensive `TextBox` instead of `TextBlock`. Your `CellEditingTemplate` is a `TextBlock` which isn't an editing control.

You can check https://github.com/AvaloniaUI/Avalonia/blob/6ca040b3b3045d98dd1a38304ed219ef737fa54d/samples/ControlCatalog/Pages/DataGridPage.xaml#L122-L133 for a XAML example.

Besides that, you'll need to be careful about trimming due to reflection in the Release build because of the Binding.

**Reply by tiborkerenyi** [reply]
Thank you for your reply, I tried the changes you mentioned, but they didn't work. (The number of columns depends on a user"s runtime choice, so I can't use the axaml, that's too static in this case. That's why I build DataTable dynamically.)

Can you suggest someone I can contact with the DataTable and DataGrid interaction?

**Reply by stevemonaco** [reply]
Not really. I rarely work with `DataGrid` and the very little I do, I use `TreeDataGrid` instead. I suggest creating a new Discussion topic here (not an Issue) for visibility.

**Reply by tiborkerenyi** [reply]
OK, solved this morning.

The datatype in template is always RataRowView, not depends on the type of the variable you sent in DataRow.

**Comment by tiborkerenyi** [comment]
Thank for all!
